revolution_pi/gpio_pin.go

```diff
@@ -145,56 +233,61 @@ func (pin *gpioPin) PWMFreq(ctx context.Context, extra map[string]interface{}) (
+	}
 
 	b := make([]byte, 1)
-	n, err := pin.ControlChip.fileHandle.ReadAt(b, 112)
+	// all PWM pins use the same PWM frequency
+	n, err := pin.ControlChip.fileHandle.ReadAt(b, int64(pin.inputOffset+outputPWMFrequencyOffset))
 	if err != nil {
 		return 0, err
 	}
 	if n != 1 {
 		return 0, errors.New("unable to read PWM Frequency")
 	}
-	pin.ControlChip.logger.Infof("Current frequency: %#v", b)
+	pin.ControlChip.logger.Debugf("Current frequency step size: %#d", b)
 
-	return 0, nil
+	return stepSizeToFreq(b), nil
 }
 
-// SetPWMFreq sets the given pin to the given PWM frequency. For Raspberry Pis,
-// 0 will use a default PWM frequency of 800.
+func stepSizeToFreq(step []byte) uint {
+	// b only has one byte
+	switch step[0] {
+	case 1:
+		return 40
+	case 2:
+		return 80
+	case 4:
+		return 160
+	case 5:
+		return 200
+	case 10:
+		return 400
+	}
+	return 0
+}
+
+// SetPWMFreq sets the given pin to the given PWM frequency. For the Rev-Pi this must be configured in PiCtory.
 func (pin *gpioPin) SetPWMFreq(ctx context.Context, freqHz uint, extra map[string]interface{}) error {
 	if !pin.initialized {
 		return errors.New("pin not initialized")
 	}
 
-	return errors.New("not supported")
+	return errors.New("PWM Frequency must be set in PiCtory")
 }
 
-// disablePwm sets the bit to turn off PWM for the pin in the controller.
-func (pin *gpioPin) disablePwm() error {
-	// We actually need to enable PWM for the pin, we can't just set the PWM value
-	// Much like in Set, we can't modify a single bit using the regular read/write in file stream
-	// so we have to use the ioctl command to modify just a single bit
-	command := SPIValue{i16uAddress: 110 + (pin.Address - 70), i8uBit: pin.BitPosition, i8uValue: 0}
-	syscallErr := pin.ControlChip.ioCtl(uintptr(KB_SET_VALUE), unsafe.Pointer(&command))
-	if syscallErr != 0 {
-		return fmt.Errorf("error turning on pwm for output: %w", syscallErr)
-	}
-	pin.pwmMode = false
-	return nil
+// pins at 70 or 71 + inputOffset.
+func (pin *gpioPin) isOutputWord() bool {
+	return pin.Address == pin.outputOffset || pin.Address == pin.outputOffset+1
 }
 
-// enablePwm sets the bit to turn on PWM for the pin in the controller.
-func (pin *gpioPin) enablePwm() error {
-	// We actually need to enable PWM for the pin, we can't just set the PWM value
-	// Much like in Set, we can't modify a single bit using the regular read/write in file stream
-	// so we have to use the ioctl command to modify just a single bit
-	command := SPIValue{i16uAddress: 110 + (pin.Address - 70), i8uBit: pin.BitPosition, i8uValue: 1}
-	syscallErr := pin.ControlChip.ioCtl(uintptr(KB_SET_VALUE), unsafe.Pointer(&command))
-	if syscallErr != 0 {
-		return fmt.Errorf("error turning on pwm for output: %w", syscallErr)
-	}
+// pins with an offset of 72 to 87 + inputOffset.
+func (pin *gpioPin) isOutputPWM() bool {
+	return pin.Address > pin.outputOffset+outputWordToPWMOffset && pin.Address < pin.inputOffset+dioMemoryOffset
```

Comment:
Any chance the first is supposed to be `>=`?

The first part uses `outputOffset` but the second part uses `inputOffset`. Are you sure those are supposed to be mixed-and-matched on the same line like this?

Reply:
good catch, left this out when I was cleaning up some logic